r12545 | sewardj | 2012-04-29 13:35:37 +0200 (Sun, 29 Apr 2012) | 3 lines

Correctly parse DW_FORM_ref_addr -- its format is different in Dwarf2
vs Dwarf3 and later.  Fixes #298864.  (Tom Tromey, tromey@redhat.com)

Index: coregrind/m_debuginfo/readdwarf3.c
===================================================================
--- valgrind-3.7.0/coregrind/m_debuginfo/readdwarf3.c	(revision 12544)
+++ valgrind-3.7.0/coregrind/m_debuginfo/readdwarf3.c	(revision 12545)
@@ -1152,9 +1152,17 @@
             So for the moment we merely range-check, to see that they
             actually do specify a plausible offset within this
             object's .debug_info, and return the value unchanged.
+
+            In DWARF 2, DW_FORM_ref_addr is address-sized, but in
+            DWARF 3 and later, it is offset-sized.
          */
-         *cts = (ULong)(UWord)get_UWord(c);
-         *ctsSzB = sizeof(UWord);
+         if (cc->version == 2) {
+            *cts = (ULong)(UWord)get_UWord(c);
+            *ctsSzB = sizeof(UWord);
+         } else {
+            *cts = get_Dwarfish_UWord(c, cc->is_dw64);
+            *ctsSzB = cc->is_dw64 ? sizeof(ULong) : sizeof(UInt);
+         }
          TRACE_D3("0x%lx", (UWord)*cts);
          if (0) VG_(printf)("DW_FORM_ref_addr 0x%lx\n", (UWord)*cts);
          if (/* the following 2 are surely impossible, but ... */
